src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copiedState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        Object.assign(copiedState, extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of keysToRemove) {
+          delete copiedState[key];
+        }
+        break;
+
+      case 'clear':
+        copiedState = {};
+        break;
+
+      default:
+        return state;
+    }
+
+    stateHistory.push({ ...copiedState });
```

Comment:
[CODE STYLE] - According to the checklist, it is better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed. This is already done correctly in your code.